pubsub/README.md

```diff
@@ -0,0 +1,322 @@
+``` 
+
+You can also fetch test outputs from S3 to the local filesystem.
+To fetch everything from the bucket into `./output`:
+
+```shell
+./sync_outputs.py fetch-all ./output
+``` 
+
+Or, to fetch one or more tests from the bucket instead of everything:
+
+```shell
+./sync_outputs.py fetch --dest=./output pubsub-test-20200409-152658
+``` 
+
+You can list all the top-level directories in the S3 bucket (so you know what to fetch) using the `list` command:
+
+```shell
+./sync_outputs.py list
+``` 
+
+## Code Overview
+
+The test code all lives in the `test` directory.
+
+`main.go` is the actual entry point, but it just calls into the "real" main function, `RunSimulation`, which is defined
+in `run.go`.
+
+`params.go` contains the parameter parsing code. The `parseParams` function will return a `testParams` struct with
+all test parameters.
+
+The set of params provided to each test instance depends on which composition group they're in. The composition
+template we're using defines two groups: `publishers`, and `lurkers`. The lurkers and publishers have
+identical params with the exception of the boolean `publisher` param, which controls whether they will publish messages
+or just consume them.
+
+After parsing the params, `RunSimulation` will prepare the libp2p `Host`s, do some network setup and then
+call `runPubsubNode` to begin the test.
+
+`discovery.go` contains a `SyncDiscovery` component that uses the testground sync service to broadcast information about
+the test peers (e.g. addreses, whether they're honest, etc) with every other peer. It uses this information to connect
+nodes to each other in various topologies.
+
+The honest node implementation is in `node.go`, and there are also `node_v10.go` and `node_v11.go` files
+that allow us to target the new gossipsub v1.1 API or the old v1.0 API by setting a build tag. If the v1.0 API
+is used, the test will not produce any peer score information, since that was added in v1.1.
+
+The `tracer.go` file implements the `pubsub.EventTracer` interface to capture pubsub events and produce test metrics.
+Because the full tracer output is quite large (several GB for a few minutes of test execution with lots of nodes),
+we aggregate the trace events at runtime and spit out a json file with aggregated metrics at the end of the test.
+We also capture a filtered subset of the original traces, containing only Publish, Deliver, Graft, and Prune events.
+At the end of the test, we run [tracestat](https://github.com/libp2p/go-libp2p-pubsub-tracer/blob/master/cmd/tracestat/main.go)
+on the filtered traces to calculate the latency distribution and get a summary of publish and deliver counts.
+
+### Note about "attackers"
+
+There are some references in the code to "attacker" peers and "sybils". The attack code has been omitted from the
+initial public release, however, we plan to re-introduce it in the near future.
```

Comment:
I recommend to just fully omitting this. We want this to be the baseline test only and avoid overloading users with information.